blockchain/src/blockchain/push.rs

```diff
@@ -307,10 +307,21 @@ impl Blockchain {
         this.chain_store.set_head(&mut txn, &block_hash);
 
         if is_election_block {
-            this.chain_store.prune_epoch(
-                Policy::epoch_at(block_number).saturating_sub(MAX_EPOCHS_STORED),
-                &mut txn,
-            );
+            let epochs_stored = cmp::max(this.config.max_epochs_stored, Policy::MIN_EPOCHS_STORED);
+
+            // Calculate the epoch to be pruned. Saturate at zero.
+            let pruned_epoch = Policy::epoch_at(block_number).saturating_sub(epochs_stored);
+
+            if this.config.keep_history {
```

Comment:
```suggestion
            if !this.config.keep_history {
```
This logic seems backwards to me because I'd expect history nodes to set `keep_history`

Reply:
Same comment as above:
history nodes would have keep_history = true, and thus:

history nodes should only prune the chain store
fill nodes should prune the chain store + history store

so it is right as it currently is

Reply:
Ok, I see where the confusion is: We're deleting history for the full node for `pruned_epoch` which means that we will have history for a full node for `epochs_stored` and in that case we don't want any history.